Feature/3 로그인 & 회원가입 기능 구현 (#6 #7)

JunHo-YH opened the pull request:
#5 기능 설계 요구사항에 따른 #6 로그인 기능 구현 #7 회원가입 기능 구현

- JWT 방식으로 Stateless 로그인 설계
- MyBatis를 사용하여 데이터베이스 쿼리 CRUD
- SpringSecurity를 사용하여 인증 및 인가
  - 클라이언트에서 쿠키를 통해 서버에 전달하는 상황을 가정하여 설계
- EmailSender 라이브러리를 사용하여 인증번호 요청
  - 요청 시 메일 요청 완료 응답까지 시간이 지연되어 비동기 처리
- AccessToken 만료 시 RefreshToken 검증을 통해 재발급
  - RefreshToken 만료시간이 지나지 않아야만 재발급 가능
  - 재발급 시 Accesstoken과 RefreshToken 재발급
  - RefreshToken은 DB에 새로 등록
  - 추후 만료된 RefreshToken은 만료 상태처리 및 쿼리 삭제를 위한 Batch 또는 Scheduler 사용 필요성을 고려중
 - 이메일 인증 제한 방안 고려
    - 하루 10회 제한, 한번 요청 후 1분 후 요청 가능한 로직을 추가
       1. 이메일을 기준으로 하루 10회 제한
       2. IP를 기준으로 하루 10회 제한

그러나 두 가지 모두 문제점이 존재한다고 생각한다.
이메일을 기준으로 했을 때
steady8@kakao.com으로 요청을 10회 한다면,
이 이메일을 사용하고 싶은 다른 사용자가 11회째에 요청을 한다면 사실상 사용 가능함에도 불구하고 다른 사용자는 다른 이메일을 선택해야 한다.
IP를 기준으로 한다면,
카페에서 많은 사람들이 같은 IP를 사용했을 때, 그 IP로는 (해당 공간에서) 최대 10번까지만 요청이 가능하므로 문제가 생길 것 같다.

Commit:
feature/3 로그인 & 회원가입 기능 구현

- Issue #6 #7
- 기능구현에 필요한 의존성 추가 (build.gradle)
  - EmailSender
  - ModelMapper
  - Validation
  - JWT

Commit:
feature/3 로그인 & 회원가입 기능 구현

- Issue #6 #7
- 이메일을 통한 인증번호 발송 기능 구현
- 이메일 중복 검증 기능 구현
- 인증번호 검증 기능 구현

Commit:
feature/3 로그인 & 회원가입 기능 구현

- Issue #6 #7
- 로그인 기능 구현
  - 로그인 시 AccessToken과 RefreshToken을 발급한다.
- 회원가입 기능 구현
- RefreshToken 재발급 기능 구현
  - 만료된 토큰은 예외가 발생한다.
  - RefreshToken으로 검증 후 AccessToken과 RefreshToken을 발급한다.

Commit:
feature/3 로그인 & 회원가입 기능 구현

- Issue #6 #7
- 이메일을 통한 인증번호 발송 기능 구현
- 이메일 중복 검증 기능 구현
- 인증번호 검증 기능 구현

Commit:
feature/3 로그인 & 회원가입 기능 구현

- Issue #6 #7
- 설정 클래스 파일

Commit:
feature/3 로그인 & 회원가입 기능 구현

- Issue #6 #7
- UtilClass

Commit:
feature/3 로그인 & 회원가입 기능 구현

- Issue #6 #7
- GlobalException

Commit:
feature/3 로그인 & 회원가입 기능 구현

- Issue #6 #7
- SpringSecurity
  - 회원가입, 로그인에 필요한 API는 권한 없이 요청 가능
  - 나머지 API는 BASIC 권한 이상만 요청 가능하도록 설정
  - AccessToken 쿠키로부터 요청 받아서 검증

Commit:
feature/3 로그인 & 회원가입 기능 구현

Commit:
feature/3 로그인 & 회원가입 기능 구현
- application.yml
  - mybatis를 위한 전역 설정

Commit:
feature/3 로그인 & 회원가입 기능 구현
- 정적 리소스(css,js,images) 허용
- 루트, 헬스 체크, 에러페이지, favicon.ico 허용

f-lab-maverick commented on src/main/java/kr/gravy/gravy/configuration/SecurityConfiguration.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
```
BCryptPasswordEncoder를 사용할 때 strength, random을 커스텀하게 설정할 수 있습니다.
또한 해당 인코더는 salt를 자동으로 사용해주는데요, 해쉬화된 암호문과 salt를 같이 저장할 때 어떻게 저장할까요? 어떻게 해당 암호문에서 salt와 암호문을 분리해낼 수 있을까요?

https://docs.spring.io/spring-security/reference/api/java/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.html 문서를 확인해보세요~

JunHo-YH replied:
<img width="848" height="435" alt="image" src="https://github.com/user-attachments/assets/afca3dc2-4b85-4182-8da6-b367bb67135c" />
[BCrypt](https://ko.wikipedia.org/wiki/Bcrypt)

strength는 bcrypt의 cost이며, 값이 클수록 해시 시간이 길어지지만 그만큼 안전해집니다
기본값은 10이며 4~31 사이로 설정할 수 있습니다.

필요 시 SecureRandom 구현체를 직접 넘겨서 랜덤 시드 소스도 커스터마이즈 할 수 있습니다.

bcrypt는 버전, cost, salt, hash를 하나의 60자 문자열로 만듭니다.
형태는 첨부된 이미지와 같습니다.
```$2a$10$<22자 salt><31자 hash>```
문자열 안에 포함이 되므로 salt를 같이 저정하는 형태로 이해했습니다.
------
salt는 bcrypt 문자열 안에 암호화된 형태로 포함되어 있어 별도로 저장할 필요가 없음을 이해했습니다.
해시를 검증할 때는 이 문자열 안의 salt를 다시 꺼내 raw 값을 해시하여 비교하기 때문에 내부적으로 자동 처리됩니다.

BCrypt의 암호화 문자열에 형식이 있기 때문에 정규 표현식으로 파싱을 통해 salt와 hash 분리가 가능할 것 같습니다.

matches(raw, encoded)는 내부적으로 이 형식의 구조를 읽어서 encoded 문자열 속 salt를 사용하여 검증하는 것으로 이해했습니다.

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/controller/EmailVerificationController.java:
```diff
@@ -0,0 +1,41 @@
+
+    private final EmailVerificationService emailVerificationService;
+
+    @PostMapping("/email/verification-code/send")
```
리소스중심이 아닌 행위중심의 URL 설계가 보입니다.
```
@PostMapping("/email/verification-code/send")  // 행위 중심
@PostMapping("/email/duplicate")               // 행위 중심
```

이런식으로 변경하셔야 합니다.
```
@PostMapping("/api/v1/email-verifications")     // 이메일 인증 생성
@GetMapping("/api/v1/users/email/{email}/availability")  // 이메일 사용 가능 여부 조회
@PostMapping("/api/v1/auth/refresh")           // 토큰 갱신
```

JunHo-YH replied:
[수정 완료]
[4e531b9](https://github.com/f-lab-edu/gravy/pull/8/commits/4e531b950146dcb343ab7716f27d4ea03b0a4a7f)
아직 리소스 중심으로 설계하는 방식이 다소 어색하지만, 이 관점으로 생각할 수 있도록 노력하겠습니다 !

이 과정에서 한 가지 궁금했던 점이 있습니다.
```
@GetMapping("/api/v1/users/email/{email}/availability")  // 이메일 사용 가능 여부 조회
```
email처럼 사용자의 정보를 URL에 노출되는게 꺼려지는데, 괜찮을까요..?
찾아본 문제점으로는 브라우저 히스토리에 남고, 외부 링크로 이동할 때 Referrer 헤더로 제 3자에게 유출될 수 있는 문제가 있다고 합니다.

해결 방법으로는 Nginx 등에서 마스킹하는 방법도 있지만, 그보다는 MessageBody를 활용할 수 있는 방식이 더 쉽고 안전하다고 생각했습니다.
PUT, PATCH, DELETE는 의미상 맞지 않아서 아래처럼 액션 메서드 스타일이 가장 적절하지 않을까 생각합니다.

```
@PostMapping("/api/v1/users/email/availability:check")
```
이렇게 하면 MessageBody로 이메일을 안전하게 전달할 수 있고,
URl 설계에도 검증이라는 의미를 명확하게 전달할 수 있다고 생각합니다.

-----
질문을 정리하자면, 보안적인 이유로 리소스 중심 설계가 애매해질 때 액션 메서드 스타일을 사용하는게 괜찮은 선택인지 궁금합니다 😃

f-lab-maverick replied:
@JunHo-YH 이 요청이 실제 "개인정보"를 포함하는가가 관점인데요, 해당 api는 이 이메일의 사용 가능여부를 조회하는거라서 말씀해주신대로 가능성이 존재하네요.

이런 경우 파라미터를 BODY로 받는 경우가 많습니다. 보호조치를 하고싶다면 "추가 이슈: 이메일 파라미터 보안 취약점 개선" 정도로 이슈 만들어주시면 됩니다 :)

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/controller/EmailVerificationController.java:
```diff
@@ -0,0 +1,41 @@
+
+    private final EmailVerificationService emailVerificationService;
+
+    @PostMapping("/email/verification-code/send")
```
API URL을 설계할 때에는 API 버전 관리가 필요합니다.
API는 한번  생성해두면 파라미터, 엔드포인트 등이 변경되지 않아야 합니다. 변경이 필요할 때에는 새로운 버전을 따서 신규 엔드포인트를 만들어 제공합니다.

아래 링크를 확인하셔서 API 버전 관리 개념을 확인해보시죠~
https://www.postman.com/api-platform/api-versioning/

JunHo-YH replied:
[수정 완료]
[4e531b9](https://github.com/f-lab-edu/gravy/pull/8/commits/4e531b950146dcb343ab7716f27d4ea03b0a4a7f)

- API는 한 번 만들어지면 파라미터나 엔드포인트가 바뀌면 안 된다.
- 변경이 필요할 때는 버전 관리 전략을 사용해서 새 API 엔드포인트로 제공해야 한다.
- 변경이 언제나 새 버전의 필요성을 의미하는 것은 아니므로 호환성을 유지할 수 있는 방법이 없는지 먼저 확인해야 한다.
- 새 버전을 출시할 경우 API 문서에 변경 사유, 영향 범위, 접근법, 마이그레이션 가이드 등을 명확히 기록해야 한다.
- 한번에 전면 적용하기보다 소수 사용자 그룹부터 단계적으로 배포하고, 피드백을 받아 수정해야 안정적으로 운영할 수 있다.
- 새 버전 도입 후 적절한 시점에 구버전 사용 현황을 모니터링 하고, 더 이상 사용되지 않는다면 공식적으로 폐기 일정을 공지하여 단계적으로 종료해야 한다.

버전 관리에 대해서는 고민해보지 못했었는데, 제공해주신 문서를 통해 생각해볼 수 있는 계기가 되었습니다 ! 감사합니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/configuration/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,85 @@
+            }
+        }
+        return null;
+    }
```
HttpOnly, Secure 속성을 확인하지 않고 쿠키를 그대로 사용하고 있습니다. XSS 공격으로 토큰 탈취 위험이 발생합니다.

JunHo-YH replied:
[수정 완료]
[da4e96d](https://github.com/f-lab-edu/gravy/pull/8/commits/da4e96dbaba93229c4e1e0b1d6785d39c2bcbf89)

이번 커밋으로, 보안에 대해서 더 깊게 생각해볼 수 있는 계기가 되었습니다.

XSS, CSRF와 같은 공격에 대해 이해하고, 서버에서 토큰을 검증하여 철저히 막아야겠다는 생각이 들었습니다.

추후 모니터링, 로깅을 통해 해당 요청의 IP 벤하는 것을 처리해야 겠다는 생각도 들었습니다 !

f-lab-maverick commented on src/main/resources/mapper/EmailVerificationMapper.xml:
```diff
@@ -0,0 +1,41 @@
+    <select id="getLatestVerification" resultMap="EmailVerificationMap">
+        SELECT id, public_id, verification_code, created_at
+        FROM email_verification
+        WHERE email = #{email}
```
MyBatis에서 `#{email}`, `${email}`은 어떤 차이가 있을까요?

JunHo-YH replied:
전달받은 값에 문자열 처리 여부에 대한 차이점이 있었습니다.

email 문자열이 steady8@kakao.com이었다면
```
#{email}: 'steady8@kakao.com'
${email}: steady8@kakao.com
```
[4e531b9](https://github.com/f-lab-edu/gravy/pull/8/commits/4e531b950146dcb343ab7716f27d4ea03b0a4a7f)

의도치않게 ${email}을 사용했었는데, 
```
[2025-08-21 12:22:44] [42000][1064] You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '@kakao.com' at line 1
```

이메일 주소의 @ 기호가 SQL에서 특수문자로 인식되어 구문 에러가 발생하였습니다.
따라서 전달받은 값을 문자열로 인식할 수 있도록 #{email}로 수정하였습니다 !

이와 같은 문제로 SQL Injection 문제에 대해 이해할 수 있었습니다.
---------
- #{}는 PreparedStatement로 안전한 파라미터 바인딩을, ${}는 직접 문자열 치환을 수행한다.
- 보안과 안정성을 위해 항상 #{}를 사용하고, ${}는 테이블명이나 컬럼명 등 동적 SQL 구조가 필요할 때만 주의깊게 사용해야한다.

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/exception/Status.java:
```diff
@@ -0,0 +1,32 @@
+    VERIFICATION_CODE_MISMATCH(400, "VERIFY001", "인증번호가 일치하지 않습니다"),
+    EMAIL_NOT_VERIFIED(400, "VERIFY002", "이메일 인증을 다시 진행해주세요."),
+    EXPIRED_VERIFICATION_CODE(400, "VERIFY003", "이메일 인증이 만료된 코드입니다."),
+    EXISTS_ALREADY_EMAIL(400, "VERIFY003", "다른 이메일을 사용해주세요."),
```
에러코드 `VERIFY003`이 중복되네요~ 에러코드를 정의하신건 어떤 에러인지 구분하기 위해서인걸로 보이니, 모든 코드는 유니크하게 관리해주세요~

JunHo-YH replied:
[수정 완료]
[940fad9](https://github.com/f-lab-edu/gravy/pull/8/commits/940fad98c994c69a1dbed215fa745f22992a25e6)

꼼꼼한 피드백 감사합니다 ☺️

여러 블로그들을 참고해보면, 어떤 에러코드인지 추측이 불가능하도록 의미 없는 난수(NX3A9)처럼 사용하기도 하던데, 클라이언트가 이해할 수 있는(VERIFY = 이메일 인증 관련) 코드를 사용해도 괜찮을지 궁금합니다.

f-lab-maverick replied:
어차피 어떤 에러인지 알려주기 위함이기 때문에 별도 이슈 없습니다. 코드는 자유롭게 사용해주셔도 됩니다.

f-lab-maverick commented on src/main/java/kr/gravy/gravy/dto/ValidateDuplicateEmailDto.java:
```diff
@@ -0,0 +1,15 @@
+    public static class Request {
+
+        @NotBlank
+        private String email;
```
여기서는 `@Email` 검증을 안하는 이유가 있을까요?

JunHo-YH replied:
[클래스 삭제]
[4e531b9](https://github.com/f-lab-edu/gravy/pull/8/commits/4e531b950146dcb343ab7716f27d4ea03b0a4a7f)

@Email을 검증하는게 맞는데 제가 놓쳤습니다 ㅠㅠ
해당 커밋을 수정하면서 이 DTO가 필요없어져서 클래스 삭제하였습니다 !

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/service/EmailVerificationService.java:
```diff
@@ -0,0 +1,107 @@
+        emailVerificationMapper.insertVerificationCode(verificationCodeVO);
+
+        // TODO:: 하루에 메일 당 최대 10번 요청 가능, 한번 요청 후 30초 후 요청 가능
+        sendEmail(request.getEmail(), verificationCode);
```
sendEmail은 비동기함수로, 메일 발송에 실패해도 예외가 별도 스레드에서 발생합니다. 이러면 메일 발송에 실패해도 응답도 성공으로 내려가고, 트랜잭션도 그대로 롤백없이 커밋되지 않을까요?

JunHo-YH replied:
[수정 완료]
[1f68b70](https://github.com/f-lab-edu/gravy/pull/8/commits/1f68b706b06380a9c86e67f689d79a257e6da503)
- 메일 발송에 실패했을 경우 클라이언트에게 200 응답을 주지만, 트랜잭션 롤백이 되지 않아 DB에는 데이터가 적재된 상태로 남아있는 문제를 이해했습니다.
- email_verification 테이블에 이메일 발송 상태로 비동기 문제를 개선하는 방법으로 적용해 볼수 있었지만, 성능 개선 및 스레드에 대한 이해도를 높이며 적용해보도록 하겠습니다.
- 코드 수정한 이 커밋에서는 동기로 처리하되, 5초 이상 SMTP 연결 및 응답이 지연될 경우 MailException에 대한 예외를 처리하도록 수정하였습니다.

피드백 감사합니다 ☺️

[동기 방식에서 고민하고 있는 문제점]
- 서버의 스레드가 100개, 메일 발송 요청시 5초가 걸리는 것으로 가정했을 때
100명의 사용자가 동시에 회원가입 요청을 하면 기존에 서비스를 이용중인 사용자가 사용할 스레드가 없기 때문에 5초 동안 기다려야하는 문제점이 존재한다. (Thread pool이 고갈될 위험이 존재할 때, 이메일 발송 요청이 아닌 다른 요청을 하는 사용자에게도 영향을 미치는 문제점이 있다.)

f-lab-maverick replied:
동시성이슈는 별도 이슈로 개선 진행하시죠 :)

f-lab-maverick commented on src/main/java/kr/gravy/gravy/service/UserService.java:
```diff
@@ -0,0 +1,125 @@
+                .expiredAt(refreshTokenExpiredDate)
+                .build();
+
+        refreshTokenMapper.insertRefreshToken(refreshTokenVO);
```
기존 토큰을 만료시키지 않고 로그인할 때 마다 신규 토큰을 발급한다면 하나의 유저가 수많은 인증 refreshToken을 가질 수 있지 않을까요? 이런 경우 기존 refreshToken을 만료시켜주는게 좋습니다~

JunHo-YH replied:
[수정 완료]
[1f68b70](https://github.com/f-lab-edu/gravy/pull/8/commits/1f68b706b06380a9c86e67f689d79a257e6da503)
- refreshToken을 무효화하는 로직을 추가하였습니다.
- 만료시킬 때 refreshToken의 상태값을 기준으로 무효화(REVOKED)처리 하였습니다.
- 토큰을 검증할 때는 ACTIVE 상태를 기준으로 활성화 여부를 판단하도록 설계하였습니다.
- 추후 기간이 만료된 토큰을 삭제시켜야할 상황을 고려하여 EXPIRED 상태와 무효화(REVOKED) 상태로 구분하였습니다.

꼼꼼한 피드백 감사합니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/vo/UserVO.java:
```diff
@@ -0,0 +1,33 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserVO {
```
VO는 말 그대로 값 객체라서 비즈니스로직이나 setter등의 로직이 존재하면 안됩니다. validatePassword는 유저 도메인 또는 패스워드 인증 도메인에서 처리해야할 관심사이며, VO에 존재하면 안됩니다.

VO, DTO, Entity(Domain)의 개념을 다시 고민해보시고, 활용하는 영역에 따라 클래스를 구분해서 사용해주세요.

JunHo-YH replied:
[수정 완료]
[578f111](https://github.com/f-lab-edu/gravy/pull/8/commits/578f11138e634e9b8982385539e759a79b96a01b)

기존에 VO를 값 객체라는 개념보다는 "서버 내부에서만 사용하는 값"으로 이해하고 있었던게 문제였습니다 😭
DTO도 클라이언트에 종속된 외부 통신 전용 객체가 아니라 "계층 간 데이터 전송 객체"로서 데이터베이스와 통신하기 위해서도 사용하는 클래스임을 이해할 수 있었습니다.

제 의도는 User, RefreshToken, EmailVerification 클래스는 Entity와 의도가 맞아떨어지기 때문에 entity 패키지로 분리하여 수정하였습니다.

@Builder, resultMap에 대한 수정은 다음 커밋에서 수정하도록 하겠습니다 !

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,63 @@
+    // 에러 응답 DTO
+    @Getter
+    @AllArgsConstructor
+    public static class ErrorResponse {
```
공통 에러응답을 만들어주셨네요~ 가능하면 HTTP API의 권장 에러 핸들링 타입인 Problem Details를 사용하는건 어떠실까요? 이 부분은 필수는 아니며, 한번 검토만 해주세요~ Spring MVC에서도 Problem Details 타입을 지원합니다.

https://datatracker.ietf.org/doc/html/rfc7807

JunHo-YH replied:
[수정 완료]
[26e0795](https://github.com/f-lab-edu/gravy/pull/8/commits/26e07950d7799d7c8bc2b6d8b9aac4015836d446)
HTTP API의 권장 에러 헨들링 타입인 Problem Details를 적용하였습니다.
Swagger와 함께 사용하면 type 부분에서 에러 해결 방법을 제시할 수 있을 것 같다는 아이디어가 떠올라서 함께 적용해보았습니다 !

ProblemDetails를 구현하는 과정에서 String 문자열 덧셈이 일어나는 부분이 있었는데, 

```
@PostConstruct
    void init() {
        initFallbackUri(baseUrl);
        for (Status status : values()) {
            status.initDocUri(baseUrl);
        }
    }
```

방식을 적용하여 로딩 시점에 한 번만 초기화 해주어 문자열 덧셈을 최소화하였습니다 !
처음에는 Enum과 분리하여 GlobalException 클래스에서 캐싱 전략을 사용하려고 했는데, httpStatusCode, code, message가 ENUM이 가지고 있는 값이기 때문에 캡슐화를 고려햐여 Enum 클래스에 ProblemDetails 메서드를 만들었습니다 ☺️


```
{
    "type": "http://localhost:8080/swagger-ui/index.html?urls.primaryName=ErrorDocumentation#/ErrorDocumentation/auth002",
    "title": "Unauthorized",
    "status": 401,
    "detail": "토큰을 찾을 수 없습니다",
    "instance": "/api/v1/auth/tokens/reissue",
    "code": "auth002",
    "timestamp": "2025-08-25T18:39:54.1801"
}
```

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,63 @@
+                "VALID001",
+                "입력값 검증에 실패했습니다: " + errors,
+                LocalDateTime.now()
+        );
```
Status enum을 사용하지 않고 하드코딩된 에러코드와 메세지를 사용중이네요. 가능하면 모든 리터럴 값은 별도 상수로 관리해주시는게 좋겠습니다.

JunHo-YH replied:
[수정 완료]
[26e0795](https://github.com/f-lab-edu/gravy/pull/8/commits/26e07950d7799d7c8bc2b6d8b9aac4015836d446)

이 커밋에서 함께 수정하였습니다 !
리터럴 값은 별도 상수로 관리하도록 항상 고려하겠습니다 !
별도 상수로 관리하니까 의미 전달과 재사용성이 높아진다는 점을 이해할 수 있었습니다 !
감사합니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/service/EmailVerificationService.java:
```diff
@@ -0,0 +1,107 @@
+    public void sendEmailVerificationCode(final SendEmailVerificationCodeDto.Request request) {
+        final String verificationCode = GeneratorUtil.generateVerificationCode();
+
+        SendEmailVerificationCodeVO verificationCodeVO = SendEmailVerificationCodeVO.builder()
```
빌더패턴은 자바에서는 생각보다 그리 좋은 패턴이 아닙니다.
직접 만든 생성자를 사용하지 않고, 변수에 설정한 기본값이 무시되며, 일부 변수를 초기화하지 않아도 알 수 있는 방법(컴파일 에러 등)가 없거든요.

가능하면 생성자 / 정적 팩토리 메서드를 사용해주세요~

JunHo-YH replied:
[수정 완료]
[d0c0911](https://github.com/f-lab-edu/gravy/pull/8/commits/d0c0911e1cb9dd8f6372aa0f3c6ac079592bdf87)

이번 리팩터링을 통해 많은 것을 배웠습니다 ☺️
- @Builder 패턴의 컴파일 타임 검증 부족 문제를 정적 팩터리 메서드로 해결할 수 있었습니다.
- 빌더 패턴에서 문제가 되었던 부분을 정적 팩터리 메서드를 사용함으로써 필수 필드 누락 방지와 기본값 설정으로 런타임 시 더 안정적으로 개발할 수 있었습니다.
- 정적 팩터리 메서드를 사용하면 성능상의 이점도 가질 수 있다는 것을 알게되었습니다.
  - 하나의 인스턴스를 생성하여 재사용할 수 있으므로 생성 비용을 절약할 수 있음을 인지했습니다.
  - entity는 인스턴스별로 상태값을 독립적으로 가져야하기 때문에, 이번 경우 매번 새로운 인스턴스를 생성합니다.

createdAt, updatedAt 같은 경우 DB에 insert된 시간을 의존하는 것보다 서버에 요청 받은 시간을 의존하는 것이 더 낫겠다는 생각이 들었습니다.
따라서 서비스 레이어의 메서드에서 요청을 받았을 시점을 기점으로 시간을 생성하니 DB에 설정했던 DEFAULT CURRENT_TIMESTAMP은 방어적인 역할이 될 수 있겠다는 생각이 들었습니다.

또한, 서버 시간 기준으로 통일하여 데이터 일관성을 확보할 수 있다고 생각했습니다.

-------
그동안 빠른 구현을 위해 편의성을 선택했었지만, 안전한 서버를 어떻게 구현할지 고민하면서 정적 팩터리 메서드로 객체 생성을 엄격히 통제하는 방향으로 사고가 전환되었습니다.

앞으로는 제가 사용한 코드가 어떤 상황에 문제가 될 수 있을지 고민하면서 개발하겠습니다 !
피드백 감사합니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/utils/GeneratorUtil.java:
```diff
@@ -0,0 +1,32 @@
+@UtilityClass
+public class GeneratorUtil {
+
+    private final int VERIFICATION_CODE_LENGTH = 6;
```
상수처리 좋습니다👍

f-lab-maverick commented on src/main/java/kr/gravy/gravy/configuration/ModelMapperConfiguration.java:
```diff
@@ -0,0 +1,18 @@
+public class ModelMapperConfiguration {
+
+    @Bean
+    public ModelMapper modelMapper() {
```
ModelMapper는 왜 사용하려고 하시나요?
소스코드를 작성할 때에는 "예상 가능한" 코드를 작성하는게 중요합니다.
ModelMapper를 사용하면 비슷한 값을 가진 객체끼리 복사하는건 쉽지만, 이걸 많이 사용하게되면 변수명 바꾸는것도 못하게 될정도로 코드가 레거시화됩니다.

또한 리플렉션 기반이라 느리기도 하고, 타입안정성이 보장되지 않으며, 어떤 필드가 어떻게 매핑되는지 디버그하기 어려워 가능하면 프로젝트에서 사용하지 않는게 좋습니다.

"너무 많은걸 해줘서 편리한 유틸" 은 경계하는게 좋습니다 :)

JunHo-YH replied:
[수정 완료]
[53d4e16](https://github.com/f-lab-edu/gravy/pull/8/commits/53d4e1688c1053cb8a7005712ae96a3e8058510a)

죄송합니다 ㅠㅠ 이번에도 마찬가지로 개발 편의성을 중점으로 생각했습니다 ㅠㅠ
런타임 기준이 아닌 컴파일 기준으로 안전하게 개발하는 방향으로 생각하겠습니다 !
피드백 감사합니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/controller/UserController.java:
```diff
@@ -0,0 +1,38 @@
+    @PostMapping("/user/signup")
+    public ResponseEntity<Void> signUp(@Valid @RequestBody UserSignUpDto.Request request) {
+        userService.signUp(request);
+        return ResponseEntity.status(HttpStatus.OK).build();
```
일반적으로 회원가입, 게시글 생성 등을 할 떄에는 200도 좋지만 201 응답을 내려주기도 합니다~

f-lab-maverick commented on src/main/java/kr/gravy/gravy/dto/ReIssueAccessTokenDto.java:
```diff
@@ -0,0 +1,18 @@
+
+        private String accessToken;
+
+        private String refreshToken;
```
Response DTO는 한번 생성하면 값이  변경될 일이 없으므로 불변객체로 만드는게 관리하기 좋습니다. 따라서 이런 불변객체성 DTO는 내부 변수를 모두 final로 관리해주세요.

JunHo-YH replied:
[수정 완료]
[f24b2f1](https://github.com/f-lab-edu/gravy/pull/8/commits/f24b2f14034cd5592545f2e05729b1d2f84878c2)

한번 생성하면 값이 변경될 일이 없는 부분은 불변으로 관리하자 ! 명심하겠습니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/vo/RefreshTokenVO.java:
```diff
@@ -0,0 +1,15 @@
+
+    private Long userId;
+    private String token;
+    private Date expiredAt;
```
JWT가 내부적으로 Date 타입을 많이 사용하기는 하나, JWT 유틸 외부에서는 LocalDate등을 사용해야 합니다. Date / LocalDate 등이 혼용되어 사용되는순간 프로젝트에는 날짜타입 컨버팅 코드가 굉장히 많이 생기거든요.

JunHo-YH replied:
[수정 완료]
[578f111](https://github.com/f-lab-edu/gravy/pull/8/commits/578f11138e634e9b8982385539e759a79b96a01b)

JWT 유틸에서는 Date를 사용하고, 데이터베이스 테이블 설계 시 날짜 관련 컬럼은 DateTime으로 했기에, LocalDateTime으로 통일하였습니다 !

피드백 감사합니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/utils/GeneratorUtil.java:
```diff
@@ -0,0 +1,32 @@
+
+    private final int VERIFICATION_CODE_LENGTH = 6;
+
+    public static String generateVerificationCode() {
```
`@UtilityClass`를 붙이면 모든 메서드가 static이 되기 떄문에 별도로 static을 안붙여주어도 됩니다.

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/utils/GeneratorUtil.java:
```diff
@@ -0,0 +1,32 @@
+
+            return code.toString();
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
```
직접 만드신 GravyException을 사용하는게 적절해보이네요

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/controller/EmailVerificationController.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @PostMapping("/email/duplicate")
+    public ResponseEntity<Void> validateDuplicateEmail(
```
`ResponseEntity<Void>`를 하지 않고, 메서드 리턴값으로 void를 그대로 두어도 동일하게 200 OK 가 내려갑니다.

`public void validateDuplicateEmail(`

조금 더 간단한 방법이지만 모든 Controller에 ResponseEntity를 사용하고싶다면 이렇게 명시적으로 `ResponseEntity<Void>`를 리턴해주어도 문제는 없습니다.

JunHo-YH replied:
꼼꼼한 피드백 감사합니다 !
개인적으로 이렇게 작성하는게 코드 일관성이 있다고 생각했습니다 ㅎㅎ
ResponseEntity<Void>로 일관된 코드를 유지해보겠습니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/configuration/MailConfiguration.java:
```diff
@@ -0,0 +1,38 @@
+    private String password;
+
+    @Bean
+    public JavaMailSender setJavaMailSender() {
```
메서드명이 `setJavaMailSender`이 된다면 bean의 이름이 어떻게 설정될까요?

JunHo-YH replied:
[수정 완료]
[e5d6384](https://github.com/f-lab-edu/gravy/pull/8/commits/e5d638447c3af7902119dac03c526a27f8180788)

빈 이름이 메서드명이 그대로 빈 이름으로  설정이 되는 것으로 이해하고 있습니다.
솔직하게 말씀드리면 빈 이름이 등록되는 것은 신경쓰지 않고 작성했었습니다.

문제점을 고민해보자면, set 접두사는 세터 메서드처럼 보이기 때문에 빈 팩터리 메서드로는 오해의 소지가 있을 것 같습니다.
의도는 메일 전송 빈을 등록하는 것이기 때문에, javaMailSender로 수정하였습니다 ☺️

f-lab-maverick commented on src/main/java/kr/gravy/gravy/configuration/SecurityConfiguration.java:
```diff
@@ -0,0 +1,77 @@
+                .csrf(AbstractHttpConfigurer::disable)
+                // 공개 API는 인증 없이 접근 허용, 그 외에는 BASIC 이상 권한 필요
+                .authorizeHttpRequests(auth -> auth
+                        .requestMatchers(PathRequest.toStaticResources().atCommonLocations()).permitAll()
```
CORS 설정은 별도로 없을까요? CORS란 무엇이고, API 서버에서는 어떻게 설정해주어야 할까요?

JunHo-YH replied:
[수정 완료]
[da4e96d](https://github.com/f-lab-edu/gravy/pull/8/commits/da4e96dbaba93229c4e1e0b1d6785d39c2bcbf89)

CORS(Cross-Origin Resource Sharing)는 브라우저 환경에서 서로 다른 출처(Origin) 간의 리소스 요청을 제어하는 보안 메커니즘입니다.

제 나름대로의 해석은, A 도메인 B 도메인이 있을 때 서버가 요청을 받을 시 다른 도메인을 거쳐서 들어왔는지 확인하고, 서버에서 허용한 도메인으로부터 요청이 들어왔는지 확인 후 데이터를 응답하는 것으로 이해했습니다.
CORS 설정을 해주지 않으면, "다른 도메인을 거쳐서 요청이 들어왔는데 서버에서 반환한 데이터를 함께 공유해서 쓸래?"하는 것과 같다고 이해했습니다!
--------------------------------
다른 추측으로는, 서버에서는 어떤 요청이든 정상적으로 로직을 처리후 응답을 하지만, 브라우저에서 값을 받았을 때 "서버에서 허용한 도메인 목록"을 브라우저에서 검토 후 검열되는 프로세스인가..? 하는 생각도 들었습니다.
만약 이게 맞다면, 서버 입장에서는 불필요한 로직을 처리하게 되는데 이걸 어떻게 개선해야하는걸까..? Nginx에서 처리할 수 있나? 하는 물음표가 생겼습니다.
--------------------------------
이번 커밋에서
1. 허용할 Origin으로 로컬 프론트 서버: http://localhost:5173, 운영 환경에서는 https://gravy.kr만 허용하도록 설정했습니다.
2. 허용 메서드는 GET, POST, PUT, DELETE, PATCH, OPTIONS로 실제 사용하는 것만 허용해두었습니다.
3. 허용 헤더는 보안적으로 문제가 될 것이 예상되지 않아 "*"로 전체 허용해둔 상태입니다.
4. JWT를 쿠키 기반으로 사용하기 때문에 allowCredentials(true)로 설정해두었습니다.
5. 불필요한 OPTIONS 요청을 줄이기 위해 Preflight 캐싱을 위한 캐시 시간인 maxAge를 10분으로 설정하였습니다만, 얼마만큼으로 설정해야 할지는 아직 잘 감이 잡히지 않습니다 ㅠㅠ


-----------
2025.8.29
테스트 해본 결과, 브라우저에서 최초 요청 시 preflight 요청을 하게되는데, 그때 서버에서 어떤 도메인을 허용하지는지 확인하게 됩니다.
이때, 허용된 도메인이 아니라면 브라우저단에서 요청 자체가 되지 않는 점을 테스트하여 확인할 수 있었습니다 !

f-lab-maverick commented on src/main/resources/mapper/UserMapper.xml:
```diff
@@ -0,0 +1,31 @@
+        SELECT id, public_id, nickname, email, password, grade, created_at
+        FROM gravy_user
+        WHERE email = #{email}
+    </select>
```
resultType, resultMap을 지정해주셔야 합니다. 메서드 반환 타입으로 추론되긴 하지만 명시적으로 지정해주어야 별도 버그나 이슈가 없습니다.

JunHo-YH replied:
[수정 완료]
[1188cce](https://github.com/f-lab-edu/gravy/pull/8/commits/1188cced03a7b771c3db6aa16b2605a831bdb3a4)

- 메서드 반환 타입으로 추론되는 것에 의존하였지만, 객체, Collection 등에서 매핑이 되지 않을 수도 있다는 문제를 확인하고 수정하였습니다.
- 멘토님께서 말씀해주신 것이 resultType, resutMap을 지정하여 javaType 옵션처럼 어떤 타입으로 반환되길 기대했는지 명시하라는 의도가 맞는지 궁금합니다.


[추가 수정 사항]
기존에 refreshTokenMapper에서 User를 조회하고 있었는데, userMapper에서 조회하는 것이 각 역할에 더 맞는 선택이라고 판단되어 
```Optional<User> getUserByRefreshToken(@Param("refreshToken") String refreshToken);```
메서드의 위치를 refreshTokenMapper -> userMapper로 위치를 변경하였습니다.

f-lab-maverick commented on src/main/java/kr/gravy/gravy/common/service/EmailVerificationService.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    private void sendEmail(final String userEmail, final String verificationCode) {
+        CompletableFuture.runAsync(() -> {
```
비동기처리 시도는 좋네요~ 다만 여러문제가 있습니다.

1. 스레드 풀을 지정하지 않았고
2. 타임아웃 설정이 없으며
3. 예외발생에 대한 핸들링이 없습니다.

어떻게 해결해야 할까요?

JunHo-YH replied:
[수정 완료]
[1f68b70](https://github.com/f-lab-edu/gravy/pull/8/commits/1f68b706b06380a9c86e67f689d79a257e6da503)
이 커밋에서 동기 방식으로 작동하도록 수정하였습니다 ☺️

- 별도 스레드 풀을 지정하지 않아 ForkJoinPool을 사용하여 스레드 풀의 스레드가 고갈되어 전체 서비스에 장애가 발생될 수 있다는 점을 인지할 수 있었습니다.
- 타임아웃 설정을 하지 않아서 SMTP에 리소스가 점유되는 문제를 인지할 수 있었습니다.
- 예외발생에 대한 핸들링을 하지 않아서 문제가 생겨도 클라이언트에서는 알 수 없다는 점을 인지할 수 있었습니다.


스레드풀에 몇 개의 스레드를 사용할지, 몇 초의 타임아웃이 적당할지, 예외발생 시 어떻게 처리할지에 대해 더 학습하고 해보고 싶습니다..! 이런 기준을 정하려면 어떤 공부를 더 하면 좋을지 궁금합니다.. 😭

f-lab-maverick reviewed (changes requested):
고생하셨습니다. 리뷰 확인해서 답변남겨주세요~

1. 패키지 구조
패키지 구조를 다시 잡아보시죠. common 패키지 내부에도 controller, service 등이 존재하고, 루트 패키지에도 controller, service등이 존재합니다. 패키지를 나눌 때에는 대표적으로 두가지 방법을 사용합니다. 아래 링크를 확인해보세요.
https://medium.com/sahibinden-technology/package-by-layer-vs-package-by-feature-7e89cde2ae3a

2. record class를 사용해보세요
불변 DTO사용시에는 record class가 생각보다 편리합니다~

Commit:
feature/3 Security - CORS 설정 추가
- 서버가 신뢰할 수 있는 도메인에서만 자원 요청을 허용하도록 제어
- preflight 요청 허용

Commit:
refactor/3 패키지 구조를 Feature 단위로 분리
- auth, email, health 기능별 패키지 분리
- User 관련 클래스는 auth 패키지로 구분
- Email 관련 클래스는 email 패키지로 구분

JunHo-YH commented:
> 고생하셨습니다. 리뷰 확인해서 답변남겨주세요~
> 
> 1. 패키지 구조
>    패키지 구조를 다시 잡아보시죠. common 패키지 내부에도 controller, service 등이 존재하고, 루트 패키지에도 controller, service등이 존재합니다. 패키지를 나눌 때에는 대표적으로 두가지 방법을 사용합니다. 아래 링크를 확인해보세요.
>    https://medium.com/sahibinden-technology/package-by-layer-vs-package-by-feature-7e89cde2ae3a
> 2. record class를 사용해보세요
>    불변 DTO사용시에는 record class가 생각보다 편리합니다~

[수정 완료]
[678f640](https://github.com/f-lab-edu/gravy/pull/8/commits/678f64017aefa336897c44eb00da18e511062678)
[923b71f](https://github.com/f-lab-edu/gravy/pull/8/commits/923b71f789c07cdf98b63f92189552168155a1cb)
피드백 감사합니다 ☺️
- 패키지 구조를 Layer 단위에서 Feature 단위로 리팩터링 하여 구조를 개선했습니다.
- 알려주신 링크를 통해 패키지를 Layer 단위로 설계했을 때의 응집도가 낮아지고 결합도가 높아지는 문제점을 Feature 단위로 설계하여 응집도를 높이고 결합도를 낮추어 문제점을 개선할 수 있음을 이해할 수 있었습니다. 감사합니다 !


[f24b2f1](https://github.com/f-lab-edu/gravy/pull/8/commits/f24b2f14034cd5592545f2e05729b1d2f84878c2)
- DTO를 private 접근 제어자로만 막을 생각만 했었는데, Setter(값 변경 메서드)를 사용하지 않았을 뿐, final이 없는 경우에는 리플렉션으로 변경이 가능한 것을 이해했습니다. 
- 반대로 final 키워드를 사용하면 일반적인 리플렉션 범위에서는 값 변경이 불가능한 점을 알 수 있었습니다.
- 자바 16버전 부터 지원되는 record 클래스 타입을 사용해서 편리하게 불변 클래스를 사용할 수 있음을 이해했습니다 !

Commit:
refactor/3 DTO를 record로 변경하여 불변성 보장
- DTO는 생성 후 값 변경이 필요하지 않으므로 불변 객체로 설계
- final 키워드보다 간결한 record 클래스 적용 (Java 16 이후 지원)

Commit:
refactor/3 ENUM 클래스 패키지명 수정
- enumeration -> model

Commit:
refactor/3 URI 설계 방식 변경 행위 -> 리소스

Commit:
refactor/3 Bean 이름 수정
- setJavaMailSender -> javaMailSender

Commit:
refactor/3 Status 에러코드 중복 해결
- EXISTS_ALREADY_EMAIL 코드를 VERIFY004로 수정
- Unique해야할 code를 중복으로 사용중이던 문제 해결

Commit:
refactor/3 VO -> Entity로 변경하여 도메인 중심으로 설계
- UserVO -> User Entity (도메인 로직 포함 = validatePassword())
- SignUpVO 삭제 -> EmailVerification Entity 사용
- 날짜 타입 일관성 유지하도록 설계
  - JWT 경계에서만 Date 사용
  - LocalDateTime으로 통일
  - DateTimeUtil <-> Date 변환 유틸리티 메서드 추가

Commit:
refactor/3 빌더 패턴을 정적 팩터리 메서드로 교체
  - User, EmailVerification, RefreshToken Entity에 정적 팩터리 메서드 도입
  - @Builder 제거하고 private 생성자 + public 정적 메서드로 객체 생성 통제
  - 컴파일 타임 시 필드 누락 방지 및 기본값 설정에 안정성 확보
  - createdAt, updatedAt을 서버 시간으로 기준을 의존하도록 변경
  - DataBase의 DEFAULT CURRENT_TIMESTAMP는 예방 차원으로 설계

Commit:
refactor/3 빌더 ModelMapperConfiguration 삭제
- ModelMapper는 리플렉션 기반이라 느리고, 타입 안정성이 보장되지 않으며, 어떤 필드가 어떻게 매핑되는지 디버그하기 어렵다는 단점이 있다.

Commit:
refactor/3 resultType, resultMap으로 반환 타입을 명시한다.
- 기존에는 메서드 반환 타입을 추론하는 것에 의존하였지만, 객체, Collection 등이 있는 경우 자동 매핑이 제대로 동작하지 않을 수 있을 경우를 대비하였습니다.
- 코드를 수정하면서 User 객체를 조회할 때 refreshTokenMapper에서 조회하고 있는 문제를 발견하여, userMapper에서 조회하도록 각 역할에 맞도록 메서드 위치를 변경하였습니다. RefreshTokenMapper -> UserMapper

Commit:
refactor/3 이메일 발송 로직 비동기 -> 동기 처리
- 메일 발송에 실패했을 경우 예외가 별도 스레드에서 발생하기 때문에 응답은 성공으로 주지만, 트랜잭션 롤백이 불가능한 문제점 개선

Commit:
refactor/3 RefreshToken 무효화 처리
- 토큰 상태를 기반으로한 무효화 처리
  - ACTIVE, REVOKED, EXPIRED 상태 관리를 위한 ENUM 추가
- 토큰 재발급 시 기존 ACTIVE 토큰 무효화(REVOKED)
- 토큰 검증 시 ACTIVE 상태 및 만료 시간 검증

Commit:
refactor/3 EmailVerification 메서드 수정
- entity 사용으로 메서드명 명확화

Commit:
refactor/3 Problem Details & Swagger 적용
- API 문서화를 위해 Swagger 적용
- HTTP API의 권장 에러 핸들링 타입인 Problem Details를 사용
- Error 발생 시 해결 방법을 제시하는 페이지 제공
- Swagger 의존성 추가
- Swagger 문서에 접근할 수 있도록 Security 설정 추가
- Swagger 상수값을 관리할 클래스 추가

Commit:
refactor/3 .gitignore에 파일 추가
- application-dev.yml

Commit:
refactor/3 Configuration Properties 구조 추가
- Record 기반 Properties 클래스 생성
- Bean Validation 적용하여 실행 시 설정값 검증
- PropertiesConfiguration에서 모든 Properties 관리
- Duration 타입 사용으로 타임아웃 설정 개선
  - 단위 혼동 방지
  - 타입 안정성 고려

Commit:
refactor/3 @Value -> @ConfigurationProperties으로 변경

Commit:
refactor/3 XSS, CSRF, CORS 공격 방어 로직 추가
- SecurityConfiguration
  - CORS 설정 추가
  - 환경별 허용 도메인 설정
    - local: localhost:5173
    - dev: https://gravy.kr
  - 클라이언트에서 JS가 읽을 수 있는 정보 설정
    - Set-Cookie, Location, Content-Disposition
    - 쿠키를 읽기 위해 Set-Cookie
    - 추후 이미지 업로드를 위해 Content-Disposition 설정
  - Preflight 요청 캐싱을 위한 setMaxAge 설정
- 쿠키 응답 방식 수정
  - 기존 MessageBody로 전달하던 방식을 Header에 쿠키를 만들어서 응답하도록 수정
  - 쿠키를 만들 때 httpOnly설정을 통해 XSS 공격 시 토큰 탈취 방어
  - 환경별 secure 설정으로 운영 환경에서 HTTPS 요청만 허용
  - CSRF 방어를 위해 sameSite 설정
    - AccessToken은 LAX
    - RefreshToken은 Strict
  - 쿠키가 전송되는 Path 설정
    - AccessToken은 /api/v1 전역
    - RefreshToken은 재발급 용도로만 한정
  - maxAge는 토큰 만료시간과 동일하게 설정
- 토큰 HttpOnly, Secure 속성 검증
  - 로컬 환경에서는 HTTPS가 아니기 때문에 로컬 환경에서는 secure 검증에서 제외하도록 설계

Commit:
refactor/3 Properties -> 논리구조 수정
- Gravy Application에 관련된 내용끼리 모음

Commit:
refactor/3 쿠키 세션 사용 및 로그아웃 기능 추가
- 사용자는 로그아웃을 눌러 삭제하는 것보다 브라우저를 바로 종료하는 경우가 더 많기에 세션 쿠키를 사용하여 브라우저 종료 시 쿠키를 남기지 않도록 설계하였습니다.
- 로그아웃 시 쿠키 만료 시간(maxAge)을 0으로 설정하여 바로 만료될 수 있도록 구현하였습니다.

Commit:
refactor/3 RefreshToken 경로 수정
- 토큰 재발급 및 로그아웃에서 refresh_token 경로가 필요하므로 공통 경로인 "/api/v1/auth/tokens"로 수정합니다.

Commit:
test/3 토큰 배포된 환경에서 쿠키의 secure 검증이 가능한지 확인하기 위한 test commit

Commit:
test/3 토큰 배포된 환경에서 쿠키의 secure 검증이 가능한지 확인하기 위한 test commit

Commit:
test/3 운영환경 httpOnly - XSS 테스트 commit

Commit:
test/3 운영환경 httpOnly - XSS 테스트 commit

Commit:
test/3 운영환경 httpOnly - XSS 테스트 commit

Commit:
test/3 운영환경 httpOnly, Secure 쿠키 확인

f-lab-maverick commented on src/main/java/kr/gravy/gravy/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,82 @@
+    // 흐름: 루트(/)페이지 접근 -> 로그인 여부 확인 -> 로그인된 사용자는 대시보드 버튼 보이게, or not: 로그인 버튼과 대시보드 버튼 -> 대시보드 버튼 누를 시 로그인 페이지로 이동
+    @PostMapping("/api/v1/auth/test")
+    public ResponseEntity<Void> checkAlreadyLogin() {
+        System.out.println("----check----");
```
` System.out.println` 은 느리기때문에 가능하면 logger를 사용해주세요~

f-lab-maverick reviewed (approved):
고생하셨습니다. 머지해주셔도 좋습니다.

Commit:
feature/3 logger 사용으로 변경
